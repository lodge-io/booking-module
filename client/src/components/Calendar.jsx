
import moment from 'moment';

const { React } = window;

const utcMoment = moment.utc;

const SHORT_DAY_NAMES = ['Su', 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa'];
const MONTH_NAMES = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];

function getCurrMonth() {
  return new Date(Date.now()).getMonth();
}

function getCurrYear() {
  return new Date(Date.now()).getFullYear();
}

function getLastMonthDate(month, year) {
  return (new Date(year, month + 1, 0)).getDate();
}

function numToMonth(num) {
  return MONTH_NAMES[num];
}

function getMonthStartWeekday(month, year) {
  return new Date(year, month).getDay();
}

function createDate(year, month, day) {
  const strYear = `${year}`;
  const strMonth = month >= 10 ? `${month + 1}` : `0${month + 1}`;
  const strDay = day >= 10 ? `${day}` : `0${day}`;
  return utcMoment(`${strYear}-${strMonth}-${strDay}`);
}

const CalendarBox = styled.div`
  padding: 0.25em 0.25em;
  text-align: center;
`;
const TableD = styled.td`
  border: 1px solid #EBEBEB;
  width: 42px; height:42px;
  background-color: white;  
  &:hover{
    background-color: #00A699;
  }
`;
TableD.displayName = 'TableD';

const TableDDisabled = styled.td`
  border: 1px solid #EBEBEB;
  width: 42px; height:42px;
  color: #D8D8D8;
`;
TableDDisabled.displayName = 'TableDDisabled';

const ArrowSpan = styled.span`
  flex-grow: 1;
  text-align: center;
  border: 1px solid black;
  border-radius: 3px;
  padding-top: 3px;
  padding-bottom: 3px;
`;
const MonthSpan = styled.span`
  flex-grow: 8;
  text-align: center;
`;


const Head = styled.th`
  width: 14%;
`;
const Table = styled.table`
  width:100%;
  padding:0px;
  border-collapse: collapse;
`;
const TopRow = styled.span`
  display: flex;
  flex-direction: row;
`;
const TableHolder = styled.div`
  width: 300px;
  padding: 20px;
  border-radius: 3px;
  user-select: none;
`;

const CalendarDate = ({ year, month, date, available, inputDate, selecting }) => {
  if (available) {
    return (
      <TableD key={date}>
        <CalendarBox
          key={date}
          className="clickableDate"
          onClick={() => {
            inputDate(createDate(year, month, date));
          }
          }
        >
          {date}
        </CalendarBox>
      </TableD>
    );
  }

  return (
    <TableDDisabled key={date}>
      <CalendarBox
        key={date}
      >
        {date}
      </CalendarBox>
    </TableDDisabled>
  );

};

class Calendar extends React.Component {
  constructor(props) {
    super(props);
    this.state = { month: getCurrMonth(), year: getCurrYear() };
    this.isValidDate = this.isValidDate.bind(this);
  }

  nextMonth() {
    this.setState((prevState) => {
      let { month, year } = prevState;
      if (month === 11) {
        month = 0;
        year += 1;
      } else {
        month += 1;
      }
      return { month, year };
    });
  }

  lastMonth() {
    this.setState((prevState) => {
      let { month, year } = prevState;
      if (month === 0) {
        month = 11;
        year -= 1;
      } else {
        month -= 1;
      }
      return { month, year };
    });
  }

  isValidDate(year, month, day) {
    const { bookings, selecting } = this.props;
    const date = utcMoment(`${year}/${month + 1}/${day}`);
    const now = utcMoment().startOf('day');
    if (date - now < 0) {
      return false;
    }
    if (bookings.length === 0) {
      return true;
    }
    let i = 0;
    if (selecting === 0) {
      while (i < bookings.length && bookings[i].endDate.diff(date) <= 0) { i += 1; }
      if (i === bookings.length) { return true; }
      if (bookings[i].startDate <= date) { return false; }
    } else {
      while (i < bookings.length && bookings[i].endDate.diff(date) < 0) { i += 1; }
      if (i === bookings.length) { return true; }
      if (bookings[i].startDate < date) { return false; }
    }
    return true;
  }

  render() {
    const { year, month } = this.state;
    const { inputDate } = this.props;
    const monthName = numToMonth(month);
    let date = 1;
    const lastDay = getLastMonthDate(month, year);
    const firstDay = getMonthStartWeekday(month, year);
    const table = [];
    let emptyNum = -1;

    while (date <= lastDay) {
      const week = [];
      for (let i = 0; i < 7; i += 1) {
        if ((date === 1 && i < firstDay) || date > lastDay) {
          week.push(<td key={emptyNum} />);
          emptyNum -= 1;
        } else {
          const myDate = date;
          week.push(
            <CalendarDate
              available={(this.isValidDate(year, month, myDate))}
              year={year}
              month={month}
              date={myDate}
              inputDate={inputDate}
            />,
          );
          date += 1;
        }
      }
      table.push(week);
    }
    return (
      <TableHolder>
        <TopRow>
          <ArrowSpan className="lastMonth" onClick={() => this.lastMonth()}>&lt;</ArrowSpan>
          <MonthSpan className="calTitleSpan">
            {monthName}
            {' '}
            {year}
          </MonthSpan>
          <ArrowSpan className="nextMonth" onClick={() => this.nextMonth()}>&gt;</ArrowSpan>
        </TopRow>
        <Table>
          <tbody>
            <tr key="header">
              {SHORT_DAY_NAMES.map(val => <Head key={val}>{val}</Head>)}
            </tr>
            {table.map((week, i) => <tr key={`weekRow:${i}`}>{week.map(val => val)}</tr>)}
          </tbody>
        </Table>
      </TableHolder>
    );
  }
}

export default Calendar;
