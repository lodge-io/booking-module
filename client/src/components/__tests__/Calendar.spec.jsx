import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import Calendar from '../Calendar';

it('renders in the dom', () => {
  const res = shallow(<Calendar />).contains(2019);
  expect(res).toBeTruthy();
});

it('renders correctly', () => {
  const tree = renderer
    .create(<Calendar />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
