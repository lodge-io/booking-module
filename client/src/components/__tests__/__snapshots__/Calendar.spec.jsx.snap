// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-gzVnrw dStQk"
>
  <span
    className="sc-bZQynM gxEjMp"
  >
    <span
      className="sc-htpNat kwiLFa"
      onClick={[Function]}
    >
      &lt;
    </span>
    <span
      className="sc-bxivhb jXJJxg"
    >
      April
      2019
    </span>
    <span
      className="sc-htpNat kwiLFa"
      onClick={[Function]}
    >
      &gt;
    </span>
  </span>
  <table
    className="sc-EHOje hhiHiM"
  >
    <tr>
      <th
        className="sc-ifAKCX cQiRZG"
      >
        Su
      </th>
      <th
        className="sc-ifAKCX cQiRZG"
      >
        Mo
      </th>
      <th
        className="sc-ifAKCX cQiRZG"
      >
        Tu
      </th>
      <th
        className="sc-ifAKCX cQiRZG"
      >
        We
      </th>
      <th
        className="sc-ifAKCX cQiRZG"
      >
        Th
      </th>
      <th
        className="sc-ifAKCX cQiRZG"
      >
        Fr
      </th>
      <th
        className="sc-ifAKCX cQiRZG"
      >
        Sa
      </th>
    </tr>
    <tr>
      <td />
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          1
        </div>
      </td>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          2
        </div>
      </td>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          3
        </div>
      </td>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          4
        </div>
      </td>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          5
        </div>
      </td>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          6
        </div>
      </td>
    </tr>
    <tr>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          7
        </div>
      </td>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          8
        </div>
      </td>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          9
        </div>
      </td>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          10
        </div>
      </td>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          11
        </div>
      </td>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          12
        </div>
      </td>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          13
        </div>
      </td>
    </tr>
    <tr>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          14
        </div>
      </td>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          15
        </div>
      </td>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          16
        </div>
      </td>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          17
        </div>
      </td>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          18
        </div>
      </td>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          19
        </div>
      </td>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          20
        </div>
      </td>
    </tr>
    <tr>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          21
        </div>
      </td>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          22
        </div>
      </td>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          23
        </div>
      </td>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          24
        </div>
      </td>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          25
        </div>
      </td>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          26
        </div>
      </td>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          27
        </div>
      </td>
    </tr>
    <tr>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          28
        </div>
      </td>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          29
        </div>
      </td>
      <td
        className="sc-bwzfXH cDziqG"
      >
        <div
          className="sc-bdVaJa iUxxxf"
          onFocus={[Function]}
          onMouseOver={[Function]}
        >
          30
        </div>
      </td>
      <td />
      <td />
      <td />
      <td />
    </tr>
  </table>
</div>
`;
